staging/src/k8s.io/dynamic-resource-allocation/structured/allocator.go

```diff
@@ -739,42 +804,63 @@ func (alloc *allocator) allocateOne(r deviceIndices, allocateSubRequest bool) (b
 
 	// We already know how many devices per request are needed.
 	if r.deviceIndex >= requestData.numDevices {
+		// We have successfully allocated devices for this request.
+		alloc.incAllocationAttempt(r.claimIndex, r.requestIndex)
 		// Done with request, continue with next one. We have completed the work for
 		// the request or subrequest, so we can no longer be allocating devices for
 		// a subrequest.
-		return alloc.allocateOne(deviceIndices{claimIndex: r.claimIndex, requestIndex: r.requestIndex + 1}, false)
+		success, err := alloc.allocateOne(deviceIndices{claimIndex: r.claimIndex, requestIndex: r.requestIndex + 1}, false)
+		// We want to propagate any errAllocationResultMaxSizeExceeded to the caller. If
+		// that error is returned here, it means none of the requests/subrequests after this one
+		// could be allocated while staying within the limit on the number of devices, so there
+		// are no solution in the current request/subrequest that would work.
+		return success, err
 	}
 
 	// Before trying to allocate devices, check if allocating the devices
 	// in the current request will put us over the threshold.
-	numDevicesAfterAlloc := len(alloc.result[r.claimIndex].devices) + requestData.numDevices
+	// We can calculate this by adding the number of already allocated devices with the number
+	// of devices in the current request, and then finally subtract the deviceIndex since we
+	// don't want to double count any devices already allocated for the current request.
+	numDevicesAfterAlloc := len(alloc.result[r.claimIndex].devices) + requestData.numDevices - r.deviceIndex
```

Comment:
This looks like a bug fix. Can you split it out into a separate PR which has a unit test that fails without this fix and passes with the fix?

Reply:
This is a pretty serious bug. I didn't realize the impact of this until now. I have created a separate PR for this: https://github.com/kubernetes/kubernetes/pull/131662